File: algorithms/671. Second Minimum Node In a Binary Tree/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def findSecondMinimumValue(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def dfs(node):
            if node:
                s.add(node.val)
                dfs(node.left)
                dfs(node.right)
        s = set()
        dfs(root)
        s.remove(min(s))
        return min(s) if s else -1

# 34 / 34 test cases passed.
# Status: Accepted
# Runtime: 36 ms

File: algorithms/671. Second Minimum Node In a Binary Tree/solution.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int findSecondMinimumValue(TreeNode* root) {
        set<int> s;
        dfs(root, s);
        s.erase(s.begin());
        return s.empty() ? -1 : *s.begin();
    }

    void dfs(TreeNode * p, set<int> & s) {
        if ( p ) {
            s.insert(p->val);
            dfs(p->left, s);
            dfs(p->right, s);
        }
    }
};

// 34 / 34 test cases passed.
// Status: Accepted
// Runtime: 3 ms
